pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/PinotHelixResourceManager.java

```diff
@@ -1977,6 +1979,80 @@ private Set<String> getAllInstancesForTable(String tableNameWithType) {
+    String controllerJobResourcePath = ZKMetadataProvider.constructPropertyStorePathForControllerJob(tableNameWithType);
+    if (_propertyStore.exists(controllerJobResourcePath, AccessOption.PERSISTENT)) {
+      ZNRecord taskResourceZnRecord = _propertyStore.get(controllerJobResourcePath, null, -1);
+      return taskResourceZnRecord.getMapFields().get(taskId);
+    } else {
+      return null;
+    }
+  }
+
+  public Map<String, Map<String, String>> getAllJobsForTable(String tableNameWithType) {
+    String jobsResourcePath = ZKMetadataProvider.constructPropertyStorePathForControllerJob(tableNameWithType);
+    if (_propertyStore.exists(jobsResourcePath, AccessOption.PERSISTENT)) {
+      ZNRecord tableJobsRecord = _propertyStore.get(jobsResourcePath, null, -1);
+      return tableJobsRecord.getMapFields();
+    } else {
+      return Collections.emptyMap();
+    }
+  }
+
+  public void addNewReloadSegmentJob(String tableNameWithType, String segmentName, String jobId,
+      int numberOfMessagesSent) {
+    Map<String, String> jobMetadata = new HashMap<>();
+    jobMetadata.put(CommonConstants.ControllerJob.CONTROLLER_JOB_ID, jobId);
+    jobMetadata.put(CommonConstants.ControllerJob.CONTROLLER_JOB_TABLE_NAME_WITH_TYPE, tableNameWithType);
+    jobMetadata.put(CommonConstants.ControllerJob.CONTROLLER_JOB_TYPE, ControllerJobType.RELOAD_SEGMENT.toString());
+    jobMetadata.put(CommonConstants.ControllerJob.CONTROLLER_JOB_SUBMISSION_TIME,
+        Long.toString(System.currentTimeMillis()));
+    jobMetadata.put(CommonConstants.ControllerJob.CONTROLLER_JOB_MESSAGES_COUNT,
+        Integer.toString(numberOfMessagesSent));
+    jobMetadata.put(CommonConstants.ControllerJob.SEGMENT_RELOAD_JOB_SEGMENT_NAME, segmentName);
+    addReloadJobToZK(tableNameWithType, jobId, jobMetadata);
+  }
+
+  public void addNewReloadAllSegmentsJob(String tableNameWithType, String jobId, int numberOfMessagesSent) {
+    Map<String, String> jobMetadata = new HashMap<>();
+    jobMetadata.put(CommonConstants.ControllerJob.CONTROLLER_JOB_ID, jobId);
+    jobMetadata.put(CommonConstants.ControllerJob.CONTROLLER_JOB_TABLE_NAME_WITH_TYPE, tableNameWithType);
+    jobMetadata.put(CommonConstants.ControllerJob.CONTROLLER_JOB_TYPE,
+        ControllerJobType.RELOAD_ALL_SEGMENTS.toString());
+    jobMetadata.put(CommonConstants.ControllerJob.CONTROLLER_JOB_SUBMISSION_TIME,
+        Long.toString(System.currentTimeMillis()));
+    jobMetadata.put(CommonConstants.ControllerJob.CONTROLLER_JOB_MESSAGES_COUNT,
+        Integer.toString(numberOfMessagesSent));
+    addReloadJobToZK(tableNameWithType, jobId, jobMetadata);
+  }
+
+  private void addReloadJobToZK(String tableNameWithType, String taskId, Map<String, String> taskMetadata) {
+    String jobResourcePath = ZKMetadataProvider.constructPropertyStorePathForControllerJob(tableNameWithType);
+    ZNRecord tableJobsZnRecord;
+
+    if (_propertyStore.exists(jobResourcePath, AccessOption.PERSISTENT)) {
```

Comment:
No need to check if the record exists. We can directly `get()` and check if the return is `null`. It can also prevent race condition.
When the record exist, we should track the record version (passing a `Stat` into the `get()`), and check the expected version when setting the record to prevent race condition

Reply:
Ack. Should we be retrying in case of a version mismatch? For now I've not added retries, in line with other version based _propertyStore updates.